core/ts.cpp

```diff
@@ -0,0 +1,237 @@
+  }
+
+  if (!no_next(val))
+  {
+    throw CosaException("Got next state variable in RHS of functional assignment");
+  }
+
+  state_updates_[state] = val;
+  trans_ = solver_->make_term(
+                              And, trans_, solver_->make_term(Equal, next_map_.at(state), val));
+}
+
+void TransitionSystem::add_invar(const Term & constraint)
+{
+  // TODO: only check this in debug mode
+  if (only_curr(constraint)) {
+    init_ = solver_->make_term(And, init_, constraint);
+    trans_ = solver_->make_term(And, trans_, constraint);
+    // add the next-state version
+    trans_ = solver_->make_term(
+                                And, trans_, solver_->substitute(constraint, next_map_));
+  } else {
+    throw CosaException("Invariants should be over current states only.");
+  }
+}
+
+void TransitionSystem::constrain_inputs(const Term & constraint)
+{
+  if (no_next(constraint)) {
+    trans_ = solver_->make_term(And, trans_, constraint);
+  } else {
+    throw CosaException("Cannot have next-states in an input constraint.");
+  }
+}
+
+void TransitionSystem::add_constraint(const Term & constraint)
+{
+  if (only_curr(constraint)) {
+    init_ = solver_->make_term(And, init_, constraint);
+    trans_ = solver_->make_term(And, trans_, constraint);
+    // add over next states
+    trans_ = solver_->make_term(
+        And, trans_, solver_->substitute(constraint, next_map_));
+  } else if (no_next(constraint)) {
+    trans_ = solver_->make_term(And, trans_, constraint);
+  } else {
+    throw CosaException("Constraint cannot have next states");
+  }
+}
+
+void TransitionSystem::name_term(const string name, const Term & t)
+{
+  if (named_terms_.find(name) != named_terms_.end()) {
+    throw CosaException("Name has already been used.");
+  }
+  named_terms_[name] = t;
+}
+
+Term TransitionSystem::make_input(const string name, const Sort & sort)
+{
+  Term input = solver_->make_symbol(name, sort);
+  inputs_.insert(input);
+  return input;
+}
+
+Term TransitionSystem::make_state(const string name, const Sort & sort)
+{
+  Term state = solver_->make_symbol(name, sort);
+  Term next_state = solver_->make_symbol(name + ".next", sort);
+  states_.insert(state);
+  next_states_.insert(next_state);
+  next_map_[state] = next_state;
+  curr_map_[next_state] = state;
+  return state;
+}
+
+Term TransitionSystem::curr(const Term & term) const
+{
+  return solver_->substitute(term, curr_map_);
+}
+
+Term TransitionSystem::next(const Term & term) const
+{
+  if (next_map_.find(term) != next_map_.end()) {
+    return next_map_.at(term);
+  }
+  return solver_->substitute(term, next_map_);
+}
+
+bool TransitionSystem::is_curr_var(const Term & sv) const
+{
+  return (states_.find(sv) != states_.end());
+}
+
+bool TransitionSystem::is_next_var(const Term & sv) const
+{
+  return (next_states_.find(sv) != next_states_.end());
+}
+
+// protected methods
+
+bool TransitionSystem::only_curr(const Term & term) const
```

Comment:
maybe we can create a common function contains(const Term & term, TermVec vec). 
only_curr(term t) := contains(t, states_)
no_next(term t) := contains(t, states_ + inputs_)

Reply:
Good idea! Implemented with b78fe5a29c59031d8c97619c21db3864ef2e1363